README.md
![travis build status](	https://img.shields.io/travis/Dinnerbone/mcstatus/master.svg)
![current version](https://img.shields.io/pypi/v/mcstatus.svg)
![supported python versions](https://img.shields.io/pypi/pyversions/mcstatus.svg)

mcstatus
========

`mcstatus` provides an easy way to query Minecraft servers for any information they can expose.
It provides three modes of access (`query`, `status` and `ping`), the differences of which are listed below in usage.

Usage
-----

```python
from mcstatus import MinecraftServer

# If you know the host and port, you may skip this and use MinecraftServer("example.org", 1234)
server = MinecraftServer.lookup("example.org:1234")

# 'status' is supported by all Minecraft servers that are version 1.7 or higher.
status = server.status()
print("The server has {0} players and replied in {1} ms".format(status.players.online, status.latency))

# 'ping' is supported by all Minecraft servers that are version 1.7 or higher.
# It is included in a 'status' call, but is exposed separate if you do not require the additional info.
latency = server.ping()
print("The server replied in {0} ms".format(latency))

# 'query' has to be enabled in a servers' server.properties file.
# It may give more information than a ping, such as a full player list or mod information.
query = server.query()
print("The server has the following players online: {0}".format(", ".join(query.players.names)))
```

Command Line Interface
```
$ mcstatus
Usage: mcstatus [OPTIONS] ADDRESS COMMAND [ARGS]...

  mcstatus provides an easy way to query Minecraft servers for any
  information they can expose. It provides three modes of access: query,
  status, and ping.

  Examples:

  $ mcstatus example.org ping
  21.120ms

  $ mcstatus example.org:1234 ping
  159.903ms

  $ mcstatus example.org status
  version: v1.8.8 (protocol 47)
  description: "A Minecraft Server"
  players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']

  $ mcstatus example.org query
  host: 93.148.216.34:25565
  software: v1.8.8 vanilla
  plugins: []
  motd: "A Minecraft Server"
  players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']

Options:
  -h, --help  Show this message and exit.

Commands:
  json    combination of several other commands with json formatting
  ping    prints server latency
  query   detailed server information
  status  basic server information
```

Installation
------------

mcstatus is available on pypi, and can be installed trivially with:

```bash
python3 -m pip install mcstatus
```

Alternatively, just clone this repo!

License
-------

mcstatus is licensed under Apache 2.0.


CHANGELOG
3.1
    * Added IPv6 support

3.0
    * Drop support for python 2 and python 3.4; you may use version 2.3 or prior for outdated python versions.

2.3
    * Added json CLI command

2.2
    * Added new CLI
    * Clean up sockets better
    * Fixed connecting to ports greater than 32767
    * Took a very very long time. :D

2.1
    * Fixed the connections hanging when no data is received
    * Sped up Query requests by caching DNS
    * Fixed running setup.py when Six is not installed

2.0
    * Rewrote all the things!
    * Removed CLI. May add it back in future, if there's demand.
    * Supports query and status.
    * Supports SRV lookups.
    * Many tests! Python 3 support! Extra hugs!

1.0

    * Initial release.


mcstatus/scripts/mcstatus.py
import click
from json import dumps as json_dumps

from .. import MinecraftServer

server = None


@click.group(context_settings=dict(help_option_names=['-h', '--help']))
@click.argument("address")
def cli(address):
    """
    mcstatus provides an easy way to query Minecraft servers for
    any information they can expose. It provides three modes of
    access: query, status, and ping.

    Examples:

    \b
    $ mcstatus example.org ping
    21.120ms

    \b
    $ mcstatus example.org:1234 ping
    159.903ms

    \b
    $ mcstatus example.org status
    version: v1.8.8 (protocol 47)
    description: "A Minecraft Server"
    players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']

    \b
    $ mcstatus example.org query
    host: 93.148.216.34:25565
    software: v1.8.8 vanilla
    plugins: []
    motd: "A Minecraft Server"
    players: 1/20 ['Dinnerbone (61699b2e-d327-4a01-9f1e-0ea8c3f06bc6)']
    """
    global server
    server = MinecraftServer.lookup(address)


@cli.command(short_help="prints server latency")
def ping():
    """
    Ping server for latency.
    """
    click.echo("{}ms".format(server.ping()))


@cli.command(short_help="basic server information")
def status():
    """
    Prints server status. Supported by all Minecraft
    servers that are version 1.7 or higher.
    """
    response = server.status()
    click.echo("version: v{} (protocol {})".format(response.version.name.encode("utf-8"), response.version.protocol))
    click.echo("description: \"{}\"".format(response.description))
    click.echo(
        "players: {}/{} {}".format(
            response.players.online,
            response.players.max,
            [
                "{} ({})".format(player.name.encode("utf-8"), player.id)
                for player in response.players.sample
            ] if response.players.sample is not None else "No players online"
        )
    )


@cli.command(short_help="all available server information in json")
def json():
    """
    Prints server status and query in json. Supported by all Minecraft
    servers that are version 1.7 or higher.
    """
    data = {'online': False}
    # Build data with responses and quit on exception
    try:
        ping_res = server.ping()
        data['online'] = True
        data['ping'] = ping_res

        status_res = server.status(retries=1)
        data['version'] = status_res.version.name
        data['protocol'] = status_res.version.protocol
        data['motd'] = status_res.description
        data['player_count'] = status_res.players.online
        data['player_max'] = status_res.players.max
        data['players'] = []
        if status_res.players.sample is not None:
            data['players'] = [{'name': player.name, 'id': player.id} for player in status_res.players.sample]

        query_res = server.query(retries=1)
        data['host_ip'] = query_res.raw['hostip']
        data['host_port'] = query_res.raw['hostport']
        data['map'] = query_res.map
        data['plugins'] = query_res.software.plugins
    except:
        pass
    click.echo(json_dumps(data))


@cli.command(short_help="detailed server information")
def query():
    """
    Prints detailed server information. Must be enabled in
    servers' server.properties file.
    """
    response = server.query()
    click.echo("host: {}:{}".format(response.raw['hostip'], response.raw['hostport']))
    click.echo("software: v{} {}".format(response.software.version, response.software.brand))
    click.echo("plugins: {}".format(response.software.plugins))
    click.echo("motd: \"{}\"".format(response.motd.encode("utf-8")))
    click.echo(
        "players: {}/{} {}".format(
            response.players.online,
            response.players.max,
            response.players.names,
        )
    )


if __name__ == "__main__":
    cli()
